wp-content/themes/core/components/image/Controller.php

```diff
@@ -6,120 +6,216 @@
- * @property bool        $auto_sizes_attr
- * @property int         $expand
- * @property string      $html
- * @property string[]    $img_classes
- * @property string[]    $img_attrs
- * @property string      $img_alt_text
- * @property string      $link_url
- * @property string[]    $link_classes
- * @property string      $link_target
- * @property string      $link_title
- * @property string []   $link_attrs
- * @property string      $parent_fit
- * @property string      $shim
- * @property bool        $src
- * @property string      $src_size
- * @property string[]    $srcset_sizes
- * @property string      $srcset_sizes_attr
- * @property bool        $use_hw_attr
- * @property bool        $use_lazyload
- * @property bool        $use_srcset
- * @property string[]    $wrapper_attrs
- * @property string[]    $wrapper_classes
- * @property string      $wrapper_tag
- */
 class Controller extends Abstract_Controller {
+	/**
+	 * @var object
+	 */
 	public $attachment;
+	/**
+	 * @var string
+	 */
 	private $img_url;
+	/**
+	 * @var bool
+	 */
 	private $as_bg;
+	/**
+	 * @var bool
+	 */
 	private $auto_shim;
+	/**
+	 * @var bool
+	 */
 	private $auto_sizes_attr;
+	/**
+	 * @var int
+	 */
 	private $expand;
+	/**
+	 * @var string
+	 */
 	public $html;
+	/**
+	 * @var string[]
+	 */
 	private $img_classes;
+	/**
+	 * @var string[]
+	 */
 	private $img_attrs;
+	/**
+	 * @var string
+	 */
 	private $img_alt_text;
+	/**
+	 * @var string
+	 */
 	public $link_url;
+	/**
+	 * @var string[]
+	 */
 	private $link_classes;
+	/**
+	 * @var string
+	 */
 	private $link_target;
+	/**
+	 * @var string
+	 */
 	private $link_title;
+	/**
+	 * @var string[]
+	 */
 	private $link_attrs;
+	/**
+	 * @var string
+	 */
 	private $parent_fit;
+	/**
+	 * @var string
+	 */
 	private $shim;
+	/**
+	 * @var bool
+	 */
 	private $src;
+	/**
+	 * @var string
+	 */
 	private $src_size;
+	/**
+	 * @var array
+	 */
 	private $srcset_sizes;
+	/**
+	 * @var string
+	 */
 	private $srcset_sizes_attr;
+	/**
+	 * @var bool
+	 */
 	private $use_hw_attr;
+	/**
+	 * @var bool
+	 */
 	private $use_lazyload;
+	/**
+	 * @var bool
+	 */
 	private $use_srcset;
-	private $wrapper_attrs;
-	private $wrapper_classes;
+	/**
+	 * @var string[]
+	 */
+	private $attrs;
+	/**
+	 * @var string[]
+	 */
+	private $classes;
+	/**
+	 * @var string
+	 */
 	private $wrapper_tag;
 
 	public function __construct( array $args = [] ) {
+		$args = wp_parse_args( $args, $this->defaults() );
+
+		foreach ( $this->required() as $key => $value ) {
+			$args[$key] = array_merge( $args[$key], $value );
+		}
+
 		// the WordPress attachment to use - takes precedence over IMG_URL
-		$this->attachment = $args['attachment'] ?? null;
+		$this->attachment = $args['attachment'];
+		$this->attrs = (array) $args['attrs'];
+		// pass classes for figure wrapper. If as_bg is set true gets auto class of "lazyload"
+		$this->classes = (array) $args['classes'];
+		// html tag for the wrapper/background image container
```

Comment:
All these comments probably belong on defaults or in the properties themselves. What do you think?

Reply:
Good call, will put on the properties I think